File: tools/margin_sweep_analyze.py
#!/usr/bin/env python3
"""Analyze a CNET_MARGIN_SWEEP TSV: choose the v2 yield lever.

Reproduces the acquire oracle-evidence gate offline (usable/attempts >=
evidence_threshold AND usable >= min_evidence) for BOTH abstention
criteria — ordered top-3 decision margin (today's semantics) and top-3
SET margin (CNET_TOPK_SET) — across an epsilon grid, and validates the
ordered@campaign-eps prediction against the actual campaign ledger.

Usage:
  margin_sweep_analyze.py <sweep.tsv> [--ledger <base.gaps.txt>]
                          [--eps 0.02] [--evidence 0.9] [--min-evidence 16]

Evidence label: this predicts the ORACLE-EVIDENCE gate only. A predicted-
minable unit can still defer downstream (certify_failed, class_imbalance);
the campaign measured those at ~2% of attempts, so the prediction is a
tight upper bound and the ledger validation quantifies the gap.
"""
import argparse
import re
import sys
from collections import defaultdict


def read_sweep(path):
    per_unit = defaultdict(list)  # token -> [(ordered, set), ...]
    header = None
    with open(path) as f:
        for line in f:
            if line.startswith("#"):
                header = line.strip()
                continue
            if line.startswith("unit_idx"):
                continue
            p = line.split()
            if len(p) != 6:
                continue
            per_unit[int(p[1])].append((float(p[4]), float(p[5])))
    return header, per_unit


def read_ledger(path):
    """token -> 'acquired' | defer reason

    Acquired detection must cover BOTH ledger generations: v2+ rows carry
    the minted unit name (" acq_<tag>"); v1 rows have no unit column and an
    acquired row simply ends at an empty defer-reason ("... cce_cond_next -").
    """
    out = {}
    with open(path) as f:
        for line in f:
            m = re.search(r"\btk(\d+)q\d+\b", line)
            if not m:
                continue
            tok = int(m.group(1))
            if " acq_" in line or line.rstrip().endswith(" -"):
                out[tok] = "acquired"
            else:
                cols = line.split()
                # defer reason sits after the oracle name column
                reason = "deferred"
                for r in ("oracle_unfit", "insufficient_exemplars",
                          "certify_failed", "class_imbalance",
                          "unbounded_domain", "incumbent_healthy"):
                    if r in cols:
                        reason = r
                        break
                out[tok] = reason
    return out


def gate(margins, idx, eps, evidence, min_evidence):
    n = len(margins)
    usable = sum(1 for m in margins if m[idx] >= eps)
    if usable == 0:
        return "oracle_unfit", usable
    if usable < min_evidence:
        return "insufficient_exemplars", usable
    if usable / n < evidence:
        return "oracle_unfit", usable
    return "minable", usable


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("tsv")
    ap.add_argument("--ledger")
    ap.add_argument("--eps", type=float, default=0.02)
    ap.add_argument("--evidence", type=float, default=0.9)
    ap.add_argument("--min-evidence", type=int, default=16)
    ap.add_argument("--semantics", choices=("set", "ordered"), default="set",
                    help="which margin gate to use for --export-minable")
    ap.add_argument("--export-minable", metavar="PATH",
                    help="write one predicted-minable token id per line "
                         "(for CNET_UNIT_ALLOWLIST)")
    ap.add_argument("--export-unfit", metavar="PATH",
                    help="write predicted-unfit token ids (diagnostics)")
    a = ap.parse_args()

    header, per_unit = read_sweep(a.tsv)
    if not per_unit:
        sys.exit("no sweep rows in %s" % a.tsv)
    print(header or "(no header)")
    # completeness: the acquire gate divides by attempts == V (the full
    # window). A partial unit (truncated / still-being-written TSV) would
    # gate over a window-ordered prefix — drop it loudly instead.
    full = max(len(v) for v in per_unit.values())
    partial = [t for t, v in per_unit.items() if len(v) != full]
    for t in partial:
        del per_unit[t]
    if partial:
        print("WARNING: dropped %d partial unit(s) (< %d probes): %s"
              % (len(partial), full,
                 ", ".join("tk%d" % t for t in sorted(partial)[:8])))
    if not per_unit:
        sys.exit("no complete units")
    nu = len(per_unit)
    nprobe = full
    print("units=%d probes/unit=%d" % (nu, nprobe))

    # eps sweep table
    grid = [0.005, 0.01, 0.02, 0.03, 0.05, 0.08, 0.12]
    if a.eps not in grid:
        grid = sorted(set(grid + [a.eps]))
    print("\npredicted minable units (evidence>=%.2f, min_evidence=%d):"
          % (a.evidence, a.min_evidence))
    print("%-8s %12s %12s %10s" % ("eps", "ordered", "set", "recovered"))
    for eps in grid:
        yo = sum(1 for t in per_unit
                 if gate(per_unit[t], 0, eps, a.evidence,
                         a.min_evidence)[0] == "minable")
        ys = sum(1 for t in per_unit
                 if gate(per_unit[t], 1, eps, a.evidence,
                         a.min_evidence)[0] == "minable")
        mark = "  <-- campaign eps" if abs(eps - a.eps) < 1e-12 else ""
        print("%-8g %8d/%-3d %8d/%-3d %+10d%s"
              % (eps, yo, nu, ys, nu, ys - yo, mark))

    # per-unit verdicts at campaign eps
    rows = []
    for t in sorted(per_unit):
        vo, uo = gate(per_unit[t], 0, a.eps, a.evidence, a.min_evidence)
        vs, us = gate(per_unit[t], 1, a.eps, a.evidence, a.min_evidence)
        mean_set = sum(m[1] for m in per_unit[t]) / len(per_unit[t])
        rows.append((t, vo, uo, vs, us, mean_set))

    conv = [r for r in rows if r[1] != "minable" and r[3] == "minable"]
    lost = [r for r in rows if r[1] == "minable" and r[3] != "minable"]
    print("\nat eps=%g: set-semantics converts %d units, loses %d"
          % (a.eps, len(conv), len(lost)))

    # Export allowlist for CNET_UNIT_ALLOWLIST (quality-preserving screen).
    sem_idx = 3 if a.semantics == "set" else 1  # vs / vo field in rows
    minable_toks = sorted(r[0] for r in rows if r[sem_idx] == "minable")
    unfit_toks = sorted(r[0] for r in rows if r[sem_idx] != "minable")
    print("\nexport semantics=%s @ eps=%g: minable=%d unfit=%d"
          % (a.semantics, a.eps, len(minable_toks), len(unfit_toks)))
    if a.export_minable:
        with open(a.export_minable, "w") as f:
            f.write("# predicted-minable token ids (%s eps=%g evidence>=%.2f "
                    "min_ev=%d)\n"
                    % (a.semantics, a.eps, a.evidence, a.min_evidence))
            for t in minable_toks:
                f.write("%d\n" % t)
        print("wrote allowlist %s (%d tokens)" % (a.export_minable,
                                                  len(minable_toks)))
    if a.export_unfit:
        with open(a.export_unfit, "w") as f:
            f.write("# predicted-unfit token ids (%s eps=%g)\n"
                    % (a.semantics, a.eps))
            for t in unfit_toks:
                f.write("%d\n" % t)
        print("wrote unfit list %s (%d tokens)" % (a.export_unfit,
                                                   len(unfit_toks)))

    # decisiveness ranking (what WINDOW_SCREEN would sort by)
    rows.sort(key=lambda r: -r[5])
    print("\ntop 10 / bottom 10 by mean set margin (screen decisiveness):")
    for r in rows[:10]:
        print("  tk%-7d mean_set=%8.4f ordered=%s set=%s" % (r[0], r[5], r[1], r[3]))
    print("  ...")
    for r in rows[-10:]:
        print("  tk%-7d mean_set=%8.4f ordered=%s set=%s" % (r[0], r[5], r[1], r[3]))

    if a.ledger:
        led = read_ledger(a.ledger)
        tp = fp = fn = tn = 0
        fps, fns = [], []
        for (t, vo, uo, vs, us, _) in rows:
            actual = led.get(t)
            if actual is None:
                continue
            pred_min = vo == "minable"
            act_min = actual == "acquired"
            if pred_min and act_min:
                tp += 1
            elif pred_min and not act_min:
                fp += 1
                fps.append((t, actual, uo))
            elif not pred_min and act_min:
                fn += 1
                fns.append((t, uo))
            else:
                tn += 1
        print("\nvalidation vs ledger %s (ordered @ eps=%g):" % (a.ledger, a.eps))
        print("  predicted-minable & acquired:      %3d (true positive)" % tp)
        print("  predicted-minable & deferred:      %3d (downstream defers)" % fp)
        print("  predicted-unfit   & acquired:      %3d (should be ~0)" % fn)
        print("  predicted-unfit   & deferred:      %3d (true negative)" % tn)
        for t, reason, uo in fps[:12]:
            print("    fp tk%d actual=%s usable=%d" % (t, reason, uo))
        for t, uo in fns[:12]:
            print("    fn tk%d usable=%d" % (t, uo))


if __name__ == "__main__":
    main()

File: tools/margin_sweep_analyze.c
/* Analyze a CNET_MARGIN_SWEEP TSV: choose the v2 yield lever.
 *
 * Usage:
 *   margin_sweep_analyze <sweep.tsv> [--ledger PATH] [--eps 0.02]
 *     [--evidence 0.9] [--min-evidence 16] [--semantics set|ordered]
 *     [--export-minable PATH] [--export-unfit PATH]
 */
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

#define MAX_UNITS 4096
#define MAX_PROBES 4096

typedef struct {
    int token;
    int n;
    float ordered[MAX_PROBES];
    float setm[MAX_PROBES];
} Unit;

static Unit units[MAX_UNITS];
static int n_units;
static char header[1024];

static Unit *find_or_add(int tok) {
    int i;
    for (i = 0; i < n_units; i++)
        if (units[i].token == tok) return &units[i];
    if (n_units >= MAX_UNITS) return NULL;
    units[n_units].token = tok;
    units[n_units].n = 0;
    return &units[n_units++];
}

static int read_sweep(const char *path) {
    FILE *f = fopen(path, "r");
    char line[1024];
    if (!f) return -1;
    header[0] = 0;
    n_units = 0;
    while (fgets(line, sizeof line, f)) {
        int unit_idx, token, probe_idx, rank;
        float ord, setv;
        Unit *u;
        if (line[0] == '#') {
            snprintf(header, sizeof header, "%s", line);
            { size_t n = strlen(header); while (n && (header[n-1]=='\n'||header[n-1]=='\r')) header[--n]=0; }
            continue;
        }
        if (strncmp(line, "unit_idx", 8) == 0) continue;
        if (sscanf(line, "%d %d %d %d %f %f", &unit_idx, &token, &probe_idx, &rank, &ord, &setv) != 6)
            continue;
        (void)unit_idx; (void)probe_idx; (void)rank;
        u = find_or_add(token);
        if (!u || u->n >= MAX_PROBES) continue;
        u->ordered[u->n] = ord;
        u->setm[u->n] = setv;
        u->n++;
    }
    fclose(f);
    return n_units > 0 ? 0 : -1;
}

static const char *gate(const Unit *u, int idx, float eps, float evidence, int min_evidence,
                        int *usable_out) {
    int i, usable = 0;
    for (i = 0; i < u->n; i++) {
        float m = (idx == 0) ? u->ordered[i] : u->setm[i];
        if (m >= eps) usable++;
    }
    *usable_out = usable;
    if (usable == 0) return "oracle_unfit";
    if (usable < min_evidence) return "insufficient_exemplars";
    if ((float)usable / (float)u->n < evidence) return "oracle_unfit";
    return "minable";
}

typedef struct {
    int token;
    char actual[64];
} Led;

static int extract_tk(const char *line, int *tok) {
    const char *p = line;
    while ((p = strstr(p, "tk")) != NULL) {
        const char *q = p + 2;
        int a = 0;
        if (!isdigit((unsigned char)*q)) { p++; continue; }
        while (isdigit((unsigned char)*q)) a = a * 10 + (*q++ - '0');
        if (*q == 'q' && isdigit((unsigned char)q[1])) {
            *tok = a;
            return 1;
        }
        p++;
    }
    return 0;
}

static int read_ledger(const char *path, Led *led, int maxn) {
    FILE *f = fopen(path, "r");
    char line[4096];
    int n = 0;
    if (!f) return 0;
    while (fgets(line, sizeof line, f) && n < maxn) {
        int tok;
        size_t L;
        if (!extract_tk(line, &tok)) continue;
        led[n].token = tok;
        L = strlen(line);
        while (L && (line[L-1] == '\n' || line[L-1] == '\r')) line[--L] = 0;
        if (strstr(line, " acq_") || (L >= 2 && line[L-1] == '-' && line[L-2] == ' ')) {
            snprintf(led[n].actual, sizeof led[n].actual, "acquired");
        } else {
            const char *reasons[] = {
                "oracle_unfit", "insufficient_exemplars", "certify_failed",
                "class_imbalance", "unbounded_domain", "incumbent_healthy", NULL
            };
            int r;
            snprintf(led[n].actual, sizeof led[n].actual, "deferred");
            for (r = 0; reasons[r]; r++)
                if (strstr(line, reasons[r])) {
                    snprintf(led[n].actual, sizeof led[n].actual, "%s", reasons[r]);
                    break;
                }
        }
        n++;
    }
    fclose(f);
    return n;
}

static const char *led_get(Led *led, int n, int tok) {
    int i;
    for (i = 0; i < n; i++)
        if (led[i].token == tok) return led[i].actual;
    return NULL;
}

typedef struct {
    int token;
    const char *vo, *vs;
    int uo, us;
    float mean_set;
} Row;

static int cmp_token(const void *a, const void *b) {
    return ((const Row *)a)->token - ((const Row *)b)->token;
}

static int cmp_mean_desc(const void *a, const void *b) {
    float da = ((const Row *)a)->mean_set, db = ((const Row *)b)->mean_set;
    if (db > da) return 1;
    if (db < da) return -1;
    return 0;
}

int main(int argc, char **argv) {
    const char *tsv = NULL, *ledger = NULL;
    const char *export_minable = NULL, *export_unfit = NULL;
    const char *semantics = "set";
    float eps = 0.02f, evidence = 0.9f;
    int min_evidence = 16;
    int ai, i, full, nu;
    float grid[16];
    int ngrid = 0;
    Row *rows;
    int n_rows;

    for (ai = 1; ai < argc; ai++) {
        if (strcmp(argv[ai], "--ledger") == 0 && ai + 1 < argc) ledger = argv[++ai];
        else if (strcmp(argv[ai], "--eps") == 0 && ai + 1 < argc) eps = (float)atof(argv[++ai]);
        else if (strcmp(argv[ai], "--evidence") == 0 && ai + 1 < argc) evidence = (float)atof(argv[++ai]);
        else if (strcmp(argv[ai], "--min-evidence") == 0 && ai + 1 < argc) min_evidence = atoi(argv[++ai]);
        else if (strcmp(argv[ai], "--semantics") == 0 && ai + 1 < argc) semantics = argv[++ai];
        else if (strcmp(argv[ai], "--export-minable") == 0 && ai + 1 < argc) export_minable = argv[++ai];
        else if (strcmp(argv[ai], "--export-unfit") == 0 && ai + 1 < argc) export_unfit = argv[++ai];
        else if (!tsv) tsv = argv[ai];
        else {
            fprintf(stderr, "usage: %s <sweep.tsv> [options]\n", argv[0]);
            return 2;
        }
    }
    if (!tsv) {
        fprintf(stderr, "usage: %s <sweep.tsv> [options]\n", argv[0]);
        return 2;
    }
    if (read_sweep(tsv) != 0) {
        fprintf(stderr, "no sweep rows in %s\n", tsv);
        return 1;
    }
    printf("%s\n", header[0] ? header : "(no header)");

    full = 0;
    for (i = 0; i < n_units; i++)
        if (units[i].n > full) full = units[i].n;
    {
        int w = 0;
        for (i = 0; i < n_units; ) {
            if (units[i].n != full) {
                if (w < 8) {
                    if (w == 0) printf("WARNING: dropped partial unit(s)");
                    printf("%s tk%d", w ? "," : ":", units[i].token);
                    w++;
                }
                units[i] = units[--n_units];
            } else i++;
        }
        if (w) printf(" (< %d probes)\n", full);
    }
    if (!n_units) {
        fprintf(stderr, "no complete units\n");
        return 1;
    }
    nu = n_units;
    printf("units=%d probes/unit=%d\n", nu, full);

    {
        float base[] = {0.005f, 0.01f, 0.02f, 0.03f, 0.05f, 0.08f, 0.12f};
        int has = 0;
        for (i = 0; i < 7; i++) grid[ngrid++] = base[i];
        for (i = 0; i < ngrid; i++) if (fabsf(grid[i] - eps) < 1e-12f) has = 1;
        if (!has) {
            grid[ngrid++] = eps;
            /* sort */
            {
                int a, b;
                for (a = 0; a < ngrid; a++)
                    for (b = a + 1; b < ngrid; b++)
                        if (grid[b] < grid[a]) {
                            float t = grid[a]; grid[a] = grid[b]; grid[b] = t;
                        }
            }
        }
    }

    printf("\npredicted minable units (evidence>=%.2f, min_evidence=%d):\n",
           evidence, min_evidence);
    printf("%-8s %12s %12s %10s\n", "eps", "ordered", "set", "recovered");
    for (i = 0; i < ngrid; i++) {
        int yo = 0, ys = 0, u;
        float e = grid[i];
        for (u = 0; u < n_units; u++) {
            int uu;
            if (strcmp(gate(&units[u], 0, e, evidence, min_evidence, &uu), "minable") == 0) yo++;
            if (strcmp(gate(&units[u], 1, e, evidence, min_evidence, &uu), "minable") == 0) ys++;
        }
        printf("%-8g %8d/%-3d %8d/%-3d %+10d%s\n",
               e, yo, nu, ys, nu, ys - yo,
               fabsf(e - eps) < 1e-12f ? "  <-- campaign eps" : "");
    }

    rows = (Row *)calloc((size_t)n_units, sizeof(Row));
    n_rows = n_units;
    for (i = 0; i < n_units; i++) {
        int j;
        float sum = 0;
        rows[i].token = units[i].token;
        rows[i].vo = gate(&units[i], 0, eps, evidence, min_evidence, &rows[i].uo);
        rows[i].vs = gate(&units[i], 1, eps, evidence, min_evidence, &rows[i].us);
        for (j = 0; j < units[i].n; j++) sum += units[i].setm[j];
        rows[i].mean_set = sum / (float)units[i].n;
    }
    qsort(rows, (size_t)n_rows, sizeof(Row), cmp_token);

    {
        int conv = 0, lost = 0;
        for (i = 0; i < n_rows; i++) {
            if (strcmp(rows[i].vo, "minable") != 0 && strcmp(rows[i].vs, "minable") == 0) conv++;
            if (strcmp(rows[i].vo, "minable") == 0 && strcmp(rows[i].vs, "minable") != 0) lost++;
        }
        printf("\nat eps=%g: set-semantics converts %d units, loses %d\n", eps, conv, lost);
    }

    {
        int nmin = 0, nunfit = 0;
        int use_set = strcmp(semantics, "set") == 0;
        for (i = 0; i < n_rows; i++) {
            const char *v = use_set ? rows[i].vs : rows[i].vo;
            if (strcmp(v, "minable") == 0) nmin++;
            else nunfit++;
        }
        printf("\nexport semantics=%s @ eps=%g: minable=%d unfit=%d\n",
               semantics, eps, nmin, nunfit);
        if (export_minable) {
            FILE *f = fopen(export_minable, "w");
            if (f) {
                fprintf(f, "# predicted-minable token ids (%s eps=%g evidence>=%.2f min_ev=%d)\n",
                        semantics, eps, evidence, min_evidence);
                for (i = 0; i < n_rows; i++) {
                    const char *v = use_set ? rows[i].vs : rows[i].vo;
                    if (strcmp(v, "minable") == 0) fprintf(f, "%d\n", rows[i].token);
                }
                fclose(f);
                printf("wrote allowlist %s (%d tokens)\n", export_minable, nmin);
            }
        }
        if (export_unfit) {
            FILE *f = fopen(export_unfit, "w");
            if (f) {
                fprintf(f, "# predicted-unfit token ids (%s eps=%g)\n", semantics, eps);
                for (i = 0; i < n_rows; i++) {
                    const char *v = use_set ? rows[i].vs : rows[i].vo;
                    if (strcmp(v, "minable") != 0) fprintf(f, "%d\n", rows[i].token);
                }
                fclose(f);
                printf("wrote unfit list %s (%d tokens)\n", export_unfit, nunfit);
            }
        }
    }

    qsort(rows, (size_t)n_rows, sizeof(Row), cmp_mean_desc);
    printf("\ntop 10 / bottom 10 by mean set margin (screen decisiveness):\n");
    for (i = 0; i < n_rows && i < 10; i++)
        printf("  tk%-7d mean_set=%8.4f ordered=%s set=%s\n",
               rows[i].token, rows[i].mean_set, rows[i].vo, rows[i].vs);
    printf("  ...\n");
    for (i = (n_rows > 10 ? n_rows - 10 : 0); i < n_rows; i++)
        printf("  tk%-7d mean_set=%8.4f ordered=%s set=%s\n",
               rows[i].token, rows[i].mean_set, rows[i].vo, rows[i].vs);

    if (ledger) {
        Led led[8192];
        int nled = read_ledger(ledger, led, 8192);
        int tp = 0, fp = 0, fn = 0, tn = 0;
        printf("\nvalidation vs ledger %s (ordered @ eps=%g):\n", ledger, eps);
        for (i = 0; i < n_rows; i++) {
            const char *actual = led_get(led, nled, rows[i].token);
            int pred_min, act_min;
            if (!actual) continue;
            pred_min = strcmp(rows[i].vo, "minable") == 0;
            act_min = strcmp(actual, "acquired") == 0;
            if (pred_min && act_min) tp++;
            else if (pred_min && !act_min) {
                fp++;
                if (fp <= 12)
                    printf("    fp tk%d actual=%s usable=%d\n",
                           rows[i].token, actual, rows[i].uo);
            } else if (!pred_min && act_min) {
                fn++;
                if (fn <= 12)
                    printf("    fn tk%d usable=%d\n", rows[i].token, rows[i].uo);
            } else tn++;
        }
        printf("  predicted-minable & acquired:      %3d (true positive)\n", tp);
        printf("  predicted-minable & deferred:      %3d (downstream defers)\n", fp);
        printf("  predicted-unfit   & acquired:      %3d (should be ~0)\n", fn);
        printf("  predicted-unfit   & deferred:      %3d (true negative)\n", tn);
    }

    free(rows);
    return 0;
}
